server/datastore/mysql/apple_mdm.go

```diff
@@ -2862,6 +2862,65 @@ func (ds *Datastore) GetMDMAppleDefaultSetupAssistant(ctx context.Context, teamI
 	return asst.ProfileUUID, asst.UploadedAt, nil
 }
 
+func (ds *Datastore) UpdateHostDEPAssignProfileResponses(ctx context.Context, payload *godep.ProfileResponse) error {
+	if payload == nil {
+		// caller should not ensure this does not happen
+		level.Debug(ds.logger).Log("msg", "update host dep assign profiles responses received nil payload")
+		return nil
+	}
+
+	p := fleet.ToHostDEPAssignProfileResponses(payload)
+	if len(p.Failed)+len(p.NotAccessible)+len(p.Success) != len(payload.Devices) {
+		// this should never happen unless Apple changes the response format, so we log it for future debugging
+		level.Debug(ds.logger).Log("msg", "unrecognized assign profile response for one or more devices", "devices", payload.Devices)
+		return nil
+	}
+
+	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		if err := updateHostDEPAssignProfileResponses(ctx, tx, ds.logger, p.ProfileUUID, p.Success, string(fleet.DEPAssignProfileResponseSuccess)); err != nil {
+			return err
+		}
+		if err := updateHostDEPAssignProfileResponses(ctx, tx, ds.logger, p.ProfileUUID, p.NotAccessible, string(fleet.DEPAssignProfileResponseNotAccessible)); err != nil {
+			return err
+		}
+		if err := updateHostDEPAssignProfileResponses(ctx, tx, ds.logger, p.ProfileUUID, p.Failed, string(fleet.DEPAssignProfileResponseFailed)); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+func updateHostDEPAssignProfileResponses(ctx context.Context, tx sqlx.ExtContext, logger log.Logger, profileUUID string, serials []string, status string) error {
+	if len(serials) == 0 {
+		return nil
+	}
+
+	stmt := `
+UPDATE
+	host_dep_assignments
+JOIN 
+	hosts ON id = host_id
+SET
+	profile_uuid = ?,
+	assign_profile_response = ?
+WHERE
+	hardware_serial IN (?)
+`
```

Comment:
not sure if it's worth or possible to fix right now, but this made me think: didn't we find a bug related to this? something related to using `host.id` as the primary key for this table instead of `host.hardware_serial`?

Reply:
I think it has to do with hosts being deleted, but requires some digging.